### PR TITLE
Individual states for widgets

### PR DESCRIPTION
**Pull request by Morhaus**
See #1170.

I also added a few widgets and moved things around a bit more 😇

### CODE REVIEW

File: packages/instantsearch-react/examples/history.js [hunk 1]
```diff
@@ -2,7 +2,17 @@ import {useRouterHistory} from 'react-router';
 import {createHistory} from 'history';
 import qs from 'qs';
 
+function alphabeticalSort(a, b) {
```
Comment (bobylito) [inline packages/instantsearch-react/examples/history.js]: Can [localeCompare](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) do the trick?

File: packages/instantsearch-react/package.json [hunk 2]
```diff
@@ -26,8 +26,10 @@
   },
   "dependencies": {
     "algoliasearch": "^3.16.0",
-    "algoliasearch-helper": "^2.11.1",
+    "algoliasearch-helper": "algolia/algoliasearch-helper-js#develop",
```
Comment (bobylito) [inline packages/instantsearch-react/package.json]: YOLO :)

File: packages/instantsearch-react/examples/history.js (hunk 1 above)
Comment (Morhaus) [inline packages/instantsearch-react/examples/history.js]: Oh wow. Are there cases where we might need the specific behavior of localeCompare? (sorting strings based on locale)

File: packages/instantsearch-react/package.json (hunk 2 above)
Comment (Morhaus) [inline packages/instantsearch-react/package.json]: Living on the edge!

File: packages/instantsearch-react/src/InstantSearch.js [hunk 3]
```diff
@@ -1,141 +1,329 @@
...
+    };
+  }
+  return {
+    ...location,
+    search: query ? `?${qs.stringify(query, {sort: alphabeticalSort})}` : '',
+  };
+}
+
+function formatProps(props) {
+  return props.map(prop => `\`${prop}\``).join(', ');
```
Comment (bobylito) [inline packages/instantsearch-react/src/InstantSearch.js]: Is it the same as?
```js
return '`' + props.join('`, `') + '`';
```

Probably less readable :)

File: packages/instantsearch-react/src/InstantSearch.js [hunk 4]
```diff
@@ -1,141 +1,329 @@
...
+      missingProps.push(prop);
+    }
+  }
+  if (presentProps.length !== 0 && missingProps.length !== 0) {
+    const plural = missingProps.length > 1;
+    const missingPropsStr = formatProps(missingProps);
+    const presentPropsStr = formatProps(presentProps);
+    const propName = `prop${plural ? 's' : ''}`;
+    throw new Error(
+      `You must provide ${plural ? '' : 'a '} ${missingPropsStr} ${propName} ` +
```
Comment (bobylito) [inline packages/instantsearch-react/src/InstantSearch.js]: Nice and fancy error message :)

File: packages/instantsearch-react/src/InstantSearch.js [hunk 5]
```diff
@@ -1,141 +1,329 @@
...
+    apiKey: PropTypes.string.isRequired,
+    indexName: PropTypes.string.isRequired,
+
     history: PropTypes.object,
-    createURL: PropTypes.func,
-    treshold: PropTypes.number,
-    configureState: PropTypes.func,
     urlSync: PropTypes.bool,
-    trackedParameters: PropTypes.arrayOf(PropTypes.string),
+    treshold: PropTypes.number,
```
Comment (bobylito) [inline packages/instantsearch-react/src/InstantSearch.js]: Typo: threshold

**Comment by bobylito** [comment]
Good thing I had a preview of what you were doing :D LGTM (check comments though ;) )

File: packages/instantsearch-react/examples/history.js (hunk 1 above)
Comment (bobylito) [inline packages/instantsearch-react/examples/history.js]: I don't think so :)

File: packages/instantsearch-react/src/InstantSearch.js (hunk 3 above)
Comment (Morhaus) [inline packages/instantsearch-react/src/InstantSearch.js]: Yep, it is, although ESLint would want us to do
```
return `\`${props.join('`, `')}\``;
```
Although it's a bit longer, I prefer the map/join version, it has clearer intent imho.

File: packages/instantsearch-react/src/InstantSearch.js (hunk 4 above)
Comment (Morhaus) [inline packages/instantsearch-react/src/InstantSearch.js]: I think I have an extra space in here!

File: packages/instantsearch-react/src/InstantSearch.js (hunk 5 above)
Comment (Morhaus) [inline packages/instantsearch-react/src/InstantSearch.js]: TIL, thanks!

File: packages/instantsearch-react/examples/history.js (hunk 1 above)
Comment (Morhaus) [inline packages/instantsearch-react/examples/history.js]: I'll keep `alphabeticalSort` for now.

**Pull request merged**